src/Compilers/CSharp/Portable/BoundTree/Statement.cs

```diff
@@ -146,20 +146,50 @@ partial class BoundForEachStatement : IForEachLoopStatement
 
     partial class BoundSwitchStatement : ISwitchStatement
     {
+        private static readonly ConditionalWeakTable<BoundSwitchStatement, object> s_switchSectionsMappings =
+            new ConditionalWeakTable<BoundSwitchStatement, object>();
+
         IExpression ISwitchStatement.Value => this.BoundExpression;
 
-        ImmutableArray<ICase> ISwitchStatement.Cases => this.SwitchSections.As<ICase>();
+        ImmutableArray<ICase> ISwitchStatement.Cases
+        {
+            get
+            {
+                return (ImmutableArray<ICase>) s_switchSectionsMappings.GetValue(this, 
+                    switchStatement =>
+                    {
+                        return switchStatement.SwitchSections.SelectAsArray(switchSection => (ICase)new SwitchSection(switchSection));   
+                    });
+            }
+        }
 
         protected override OperationKind StatementKind => OperationKind.SwitchStatement;
-    }
 
-    partial class BoundSwitchSection : ICase
```

Comment:
I don't think it's broken, just to avoid implementing both IStatement and IExpression.